openedx_tagging/core/tagging/models/base.py

```diff
@@ -674,40 +526,24 @@ def value(self, value: str):
         self._value = value
 
     @property
-    def tag_ref(self) -> str:
+    def is_deleted(self) -> bool:
         """
-        Returns this tag's reference string.
-
-        If tag is set, then returns its id.
-        Otherwise, returns the cached _value field.
-        """
-        return self.tag.id if self.tag else self._value
-
-    @tag_ref.setter
-    def tag_ref(self, tag_ref: str):
+        Has this Tag been deleted from the Taxonomy? If so, we preserve this
+        ObjecTag in the DB but it shouldn't be shown to the user.
         """
-        Sets the ObjectTag's Tag and/or value, depending on whether a valid Tag is found.
+        return self.taxonomy is None or (self.tag is None and not self.taxonomy.allow_free_text)
 
-        Subclasses may override this method to dynamically create Tags.
-        """
-        self.value = tag_ref
-
-        if self.taxonomy:
-            try:
-                self.tag = self.taxonomy.tag_set.get(pk=tag_ref)
-                self.value = self.tag.value
-            except (ValueError, Tag.DoesNotExist):
-                # This might be ok, e.g. if our taxonomy.allow_free_text, so we just pass through here.
-                # We rely on the caller to validate before saving.
-                pass
-
-    def is_valid(self) -> bool:
-        """
-        Returns True if this ObjectTag represents a valid taxonomy tag.
-
-        A valid ObjectTag must be linked to a Taxonomy, and be a valid tag in that taxonomy.
-        """
-        return self.taxonomy.validate_object_tag(self) if self.taxonomy else False
+    def clean(self):
+        if self.tag:
+            if self.tag.taxonomy_id != self.taxonomy_id:
+                raise ValidationError("ObjectTag's Taxonomy does not match Tag taxonomy")
+            elif self.tag.value != self._value:
+                raise ValidationError("ObjectTag's _value is out of sync with Tag.value")
+        else:
+            # Note: self.taxonomy and/or self.tag may be NULL which is OK, because it means the Tag/Taxonomy
+            # was deleted, but we still preserve this _value here in case the Taxonomy or Tag get re-created in future.
+            if self._value == "":
+                raise ValidationError("Invalid _value - empty string")
```

Comment:
Oh man! Yes, better to fix that in another PR.

Reply:
@bradenmacdonald 

> I'll have to fix that in a future PR because there are too many lint issues showing up now. 

Do you want me to submit a PR for this? I've got time, and there's another couple of minor things that should be fixed:

* [test coverage for files under `tests/` isn't 100%](https://github.com/openedx/openedx-learning/actions/runs/6398610152/job/17369107033#step:7:160); some code is now unused, and some parts of the tests aren't being run.
* [some warnings](https://github.com/openedx/openedx-learning/actions/runs/6398610152/job/17369107033#step:7:49) about the test Taxonomy subclass names -- these classes just need to be renamed.

Reply:
@pomegranited Yes please, that would be great! Thank you :)